[ADD] stock_traceability_operation

clonedagain opened the pull request:
Module to show the detail of pack operations in the traceability.
Useful in multi-location setups: the location indicated on stock moves
is less precise that the on on the operations because it does not
account for putaway strategies, manual changes, and reservations made by
the scheduler.

Work in progress:
- [x] should display total quantities in grouped view
- [x] must add access rules
- [x] must add translation template + French
- [x] could be faster
- [x] Coverall wants automatic tests
- [x] needs user testing

Commit:
[ADD] stock_traceability_operation

Module to show the detail of pack operations in the traceability.

Useful in multi-location setups: the location indicated on stock moves
is less precise that the one on the operations because it does not
account for putaway strategies, manual changes, and reservations made by
the scheduler.

clonedagain commented:
@lbellier for your information.

@lmignon, @pedrobaeza, @jgrandguillaume I think this may be interesting to you. If you know others who would be interested too, @-mention them here please.

elicoidal commented on stock_traceability_operation/README.rst:
```diff
@@ -0,0 +1,65 @@
+means (automatic reservations, automatic putaway strategies, manual entry).
+
+In order to improve the traceability, this module lets you "unfold" the
+stock moves to see the actual locations which were used in the transfer.    
```
remove 'which were'?

clonedagain replied:
Done.

clonedagain commented:
Coverall was stuck but I rebased and now it's all green. Users are Ok with the feature.
Can we merge?

JosDeGraeve commented on stock_traceability_operation/README.rst:
```diff
@@ -0,0 +1,69 @@
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/153/8.0
+
+Kown issues
```
small typo, replace by 'Known issues'

Commit:
Fix typo

clonedagain commented:
@JosDeGraeve the typo is corrected.

Commit:
Fix a missing filter in lot traceability

When we search for lot traceability we first search for the corresponding stock moves.
But whenever these moves contain extra quants, we need to filter the ones of the wrong lots.

clonedagain commented:
Fixed a bug in lot traceability

pedrobaeza commented on stock_traceability_operation/models/stock_quant.py:
```diff
@@ -0,0 +1,66 @@
+
+    @api.multi
+    def action_view_quant_history(self):
+        '''Replace the action on stock moves into an action on the report
```
Use double-quote for docstring

pedrobaeza commented on stock_traceability_operation/models/stock_quant.py:
```diff
@@ -0,0 +1,66 @@
+        if not report_ids:
+            return False
+        return {
+            'domain': "[('id', 'in', "
```
You can use here the Python list and tuple elements, so no weird `str` + `map` operation is needed.

pedrobaeza commented:
Very great module. I think that you have to add another button instead of replacing old ones, called "Improved history" or similar, and the known issue is easily solved. You can hide the old buttons if you want in the customer custom module, or include a security group that hides them if not marked.

Commit:
Changes suggested during review

Triple-double-quotes for docstrings.
Plain domain list instead of a convoluted string.

clonedagain commented:
@lbellier do you agree with @pedrobaeza that we should keep the standard button and add ours alongside?
We're not calling super so we may be breaking other modules.

clonedagain commented:
@pedrobaeza the domain is `safe_eval`ed, so it has to be a string.

Commit:
Fix domain returned to client

Must be a string to be `safe_eval`ed

pedrobaeza commented:
@clonedagain, no, it's first checked if is an instance of a string, and if so, safe_evaled, but if not, is passed as is. Try and tell me.

clonedagain commented:
@pedrobaeza ok it seems the error is in the tests then. I'll try to improve that.

gurneyalex commented:
:+1:

Commit:
Revert to returning Python domain and don't use safe_eval() in the tests.

clonedagain commented:
@pedrobaeza I've changed as you suggested (fixed the test) and rebased.

lbellier commented:
@clonedagain i agree with @pedrobaeza.

Commit:
Create new buttons and methods instead of replacing the standard ones

clonedagain commented:
 @pedrobaeza @lbellier I've changed the code as you suggested: create new buttons and actions instead of replacing the standard ones.

pedrobaeza commented on stock_traceability_operation/views/stock_production_lot_view.xml:
```diff
@@ -0,0 +1,24 @@
+    <field name="model">stock.production.lot</field>
+    <field name="inherit_id" ref="stock.view_production_lot_form"/>
+    <field name="arch" type="xml">
+        <xpath expr="//button[@name='action_traceability']" position="attributes">
```
Better not to hide the button, to allow other improvements made to standard view accessible to the user when installing this module. You can hide the button in a your customer customization module if you want to clear the screen.

pedrobaeza commented on stock_traceability_operation/views/stock_production_lot_view.xml:
```diff
@@ -0,0 +1,24 @@
+        </xpath>
+        <xpath expr="//button[@name='action_traceability']" position="after">
+            <button name="action_traceability_operation" icon="fa-arrows-v" class="oe_stat_button"
+                string="Traceability" type="object" attrs="{'invisible': [('quant_ids','=',[])]}"/>
```
Call it "Traceability (by operation)" or similar to differentiate it from the other button.

clonedagain commented:
@pedrobaeza @lbellier I've pushed a new version which keeps the standard buttons and calls the new ones "Detailed traceability". Do you agree with this one?

pedrobaeza commented:
Thanks :+1:

clonedagain commented:
Can you merge this please ?

clonedagain commented:
@nhomar can you please comment on this?

From the community list I understand you have a similar module but much more advanced, do you intend to hand it over to OCA ?

moylop260 commented:
👍
